### PR TITLE
[JENKINS-61856] Add the ability to "Delete workspace before build starts" in a declarative pipeline

### PR DESCRIPTION
**Pull request by dennisl**
This pull request contains a refactoring to extend SimpleBuildWrapper instead of BuildWrapper, which makes it possible to use the this wrapper class in a declarative pipeline.

Things left to to:
* Fix so that you can use a shorthand notation in the Jenkinsfile instead of having to know the class. I haven't found out how to do that yet.
* Add documentation for declarative pipeline usage.

Question
Even though I have implemented runPreCheckout() to return true, I have not been able to use it the way I thought I should.

This works:
```
pipeline { 
    agent any
    options {
        skipDefaultCheckout(true)
    }
    stages {
        stage('Build') {
            steps {
                wrap([$class: 'PreBuildCleanup', cleanupParameter: 'CLEANUP']) {
                    checkout scm
                    echo "Building ${env.JOB_NAME}..."
                }
            }
        }
    }
}
```

But I thought that I should be able to do this. I'm using Subversion with "Lightweight checkout" checked. This however runs the implicit default scm checkout before PreBuildCleanup.
```
pipeline { 
    agent any
    stages {
        stage('Build') {
            steps {
                wrap([$class: 'PreBuildCleanup', cleanupParameter: 'CLEANUP']) {
                    echo "Building ${env.JOB_NAME}..."
                }
            }
        }
    }
}
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -42,6 +42,42 @@ job("foo") {
 }
 ```
 
+### Descriptive pipeline
```
Comment (les2) [inline README.md:45]:
```suggestion
### Declarative pipeline
```

Jenkins documentation refers to it as "declarative pipeline", e.g., [here](https://www.jenkins.io/doc/book/pipeline/syntax/)

File: README.md [hunk 2]
```diff
@@ -4,7 +4,7 @@
 
 There is a single [step](https://jenkins.io/doc/pipeline/steps/ws-cleanup) to be used whenever workspace is allocated. 
 
-## Job DSL
+### Scripted pipeline (Job DSL)
```
Comment (basil) [inline README.md:7]: Scripted Pipeline and Job DSL are two different concepts. I suggest leaving this section as-is and adding a new section covering the use of `preBuildCleanup` with Scripted Pipeline.

File: src/main/java/hudson/plugins/ws_cleanup/PreBuildCleanup.java [hunk 3]
```diff
@@ -27,18 +32,22 @@
...
+	public PreBuildCleanup() {
+	}
 
 	@Deprecated
 	public PreBuildCleanup(List<Pattern> patterns, boolean deleteDirs, String cleanupParameter, String externalDelete) {
 		this(patterns, deleteDirs, cleanupParameter, externalDelete, false);
 	}
 
-	@DataBoundConstructor
+	@Deprecated
```
Comment (basil) [inline src/main/java/hudson/plugins/ws_cleanup/PreBuildCleanup.java:51]: Now that this constructor is deprecated, all callers in test code should be migrated to a non-deprecated equivalent. I suggest using an IDE's Call Hierarchy feature to find and update these callers.

**Review by basil — COMMENTED** [review]
Thanks for the PR! This should be ready to merge once we add some integration test coverage for `preBuildCleanup` with Scripted Pipeline in `CleanupTest`. Refer to the following tests as an example:

- `pipelineWorkspaceCleanup` / `symbolAnnotationWorkspaceCleanup`
- `pipelineWorkspaceCleanupUsingPattern` / `symbolWorkspaceCleanupAnnotationUsingPattern`
- `pipelineWorkspaceCleanupUnlessBuildFails` / `symbolAnnotationWorkspaceCleanupUnlessBuildFails`

I suggest you write a similar pair of tests covering usage of `preBuildCleanup` with Scripted Pipeline both with and without symbol annotations. Note that there is no need to write separate tests for Declarative Pipeline, as Scripted Pipeline should be sufficient to exercise your `SimpleBuildWrapper` changes and it is expected that any `SimpleBuildWrapper` that works in a Scripted Pipeline would also work in a Declarative Pipeline.

Looking forward to seeing this PR merged! Let me know if you have any further questions.

**Comment by dennisl** [comment]
Thanks for you feedback @basil !

Unfortunately there is an inconsistency in how wrappers work in regular Jenkins jobs versus how they work in a declarative pipeline. That makes the usage of the wrapper unintuitive in a declarative pipeline. See https://issues.jenkins.io/browse/JENKINS-63266

So we ended up with the following solution instead of using the patch.

```
pipeline {
    agent any
    options {
        skipDefaultCheckout(true)
    }
    stages {
        stage('Build') {
            steps {
                cleanWs()
                checkout scm
                ...
            }
        }
    }
}
```

If you agree I can make another pull request, with just a documentation update, showing how to delete the workspace before checkout.

**Comment by basil** [comment]
I concur with your assessment. A documentation update would be most welcome!

**Comment by dennisl** [comment]
Great, a documentation update is available in https://github.com/jenkinsci/ws-cleanup-plugin/pull/55
Feel free to close this issue.

**Comment by basil** [comment]
> Feel free to close this issue.

Done

**Pull request closed without merging**